storage: compact printing of key ranges

RaduBerinde opened the pull request:
The range tags can be very long and distracting. This change adds a pretty print
function for ranges which limits the number of characters printed.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9431)

<!-- Reviewable:end -->

RaduBerinde committed:
storage: compact printing of key ranges

The range tags can be very long and distracting. This change adds a pretty print
function for ranges which limits the number of characters printed.

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[keys/printer_test.go, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8kk5eIS5hAYt_RoW:-KRt8kk6feP_LSb49j10:b-v4ktmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L182)):*
> ```Go
> 
> func TestPrettyPrintRange(t *testing.T) {
> 	//  /Table/42/1222
> ```

What is this comment about? It doesn't seem to apply to the next line.

---

*[keys/printer_test.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8w3QEXHaidlhbZEY:-KRt8w3QEXHaidlhbZEZ:b8hft0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L197)):*
> ```Go
> 		{tableKey, tableKey2, 8, "/Table/…"},
> 		{tableKey, tableKey2, 15, "/Table/61/…"},
> 		{tableKey, tableKey2, 20, "/Table/61/{4-500}"},
> ```

This is nice. I was wondering if we could do something like that.

---

*[keys/printer_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt91F8D6iSYNt91TT9:-KRt91F8D6iSYNt91TTA:b-ipmin3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L198)):*
> ```Go
> 		{tableKey, tableKey2, 15, "/Table/61/…"},
> 		{tableKey, tableKey2, 20, "/Table/61/{4-500}"},
> 		{tableKey, MaxKey, 10, "/{Ta…-Max}"},
> ```

Some of these are overly terse. I realize you're testing `maxChars`, but seeing them makes me wonder if we should take an initial step of not restricting `maxChars`. 

---

*[storage/replica.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8L54y1ej66jgIc08:-KRt8L55iMLZ2KN-524z:b-tr4ral) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/storage/replica.go#L203)):*
> ```Go
> 	keys.PrettyPrintRange(&buf, roachpb.Key(inconsistentDesc.StartKey),
> 		roachpb.Key(inconsistentDesc.EndKey), 30)
> 	return buf.String()
> ```

I think `Replica.rangeDesc` is only used for log messages. Rather than formatting the value on every call to `atomicRangeDesc.String()` perhaps we should make this an `atomicString` type instead and format only once when setting `Replica.rangeDesc`. I think there would need to be some renaming. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[keys/printer_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt91F8D6iSYNt91TT9:-KRtOl0aHqwZhTtDeQgZ:bw2h2qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L198)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Some of these are overly terse. I realize you're testing `maxChars`, but seeing them makes me wonder if we should take an initial step of not restricting `maxChars`. 
</details>
This needs to do a good job on ranges of the form, for example,

/Table/51/1/5874520773009788649/"c6b21f61-68a4-4eac-8bfa-a5e7a2dc3ea8"/102765-/Table/51/1/7874807763469459666/"900e1c68-74df-4312-8acb-1e6469c0e3ed"/1380714

and so without `maxChars` this isn't really useful in practice - making the traces terser was the motivation for this change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-:-KRtQEm-9OXAxPk4YkDp:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8L54y1ej66jgIc08:-KRtlUSTF6JTe09Jw-dz:b-c9pswf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/storage/replica.go#L203)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think `Replica.rangeDesc` is only used for log messages. Rather than formatting the value on every call to `atomicRangeDesc.String()` perhaps we should make this an `atomicString` type instead and format only once when setting `Replica.rangeDesc`. I think there would need to be some renaming. 
</details>
Yes, I agree, I will work on that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
storage: generate rangeDesc string once per update

Instead of storing an atomic `*RangeDescriptor` inside `Replica` and generating
the string every time we log a message or event, we now generate the string once
when the descriptor is updated and store it as an atomic `*string`.

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[keys/printer_test.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt91F8D6iSYNt91TT9:-KRxqGenqYj7N8xM5pTQ:b-brg45r) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L198)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This needs to do a good job on ranges of the form, for example,
> 
> /Table/51/1/5874520773009788649/"c6b21f61-68a4-4eac-8bfa-a5e7a2dc3ea8"/102765-/Table/51/1/7874807763469459666/"900e1c68-74df-4312-8acb-1e6469c0e3ed"/1380714
> 
> and so without `maxChars` this isn't really useful in practice - making the traces terser was the motivation for this change.
</details>
Ok. For the most part I find the key-ranges distracting and not useful for debugging. Pretty much all I want to see is the table ID and index ID. @bdarnell might feel differently.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I also find the key ranges more noisy than helpful; repeating them in every log line is not the best way to surface this information.

tschottdorf commented:
Same here. Especially since the format of the SQL query is also in the
trace.

On Sun, Sep 18, 2016, 11:24 Ben Darnell <notifications@github.com> wrote:

> I also find the key ranges more noisy than helpful; repeating them in
> every log line is not the best way to surface this information.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9431#issuecomment-247853988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135JaBEBCfZbi2D0-YIQU03iX6SN5bks5qrVfEgaJpZM4J_X0S>
> .
>
-- 

-- Tobias

RaduBerinde commented:
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[keys/printer_test.go, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8kk5eIS5hAYt_RoW:-KRyuXcpBMYRXp_UIHMT:bu9laei) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/keys/printer_test.go#L182)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What is this comment about? It doesn't seem to apply to the next line.
</details>
Fixed, it was left over.

---

*[storage/replica.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-KRt8L54y1ej66jgIc08:-KRyudvs5RSv9vic04oc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db10f07212a04d2a31933ef220c34c81f6d3091/storage/replica.go#L203)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Yes, I agree, I will work on that.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431#-:-KRzvVlsm7_uvY6WaB8e:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9431)*
<!-- Sent from Reviewable.io -->
